src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs

```diff
@@ -5,94 +5,45 @@
 #nullable enable
 
 using System;
-using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
-using Microsoft.CodeAnalysis.Features.EmbeddedLanguages;
-using Microsoft.CodeAnalysis.Options;
-using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Host;
 
 namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     /// <summary>
-    /// The singular completion provider that will hook into completion and will
-    /// provider all completions across all embedded languages.
-    /// 
-    /// Completions for an individual language are provided by
-    /// <see cref="IEmbeddedLanguageFeatures.CompletionProvider"/>.
+    /// Base class for emebedded language completion providers.
+    /// Since <see cref="IEmbeddedLanguage"/> are provided via document services,
+    /// this class helps to initialize and retrieve the relevant <see cref="IEmbeddedLanguage"/>
+    /// for this completion provider only once for the lifetime of the completion provider.
     /// </summary>
-    internal abstract class AbstractEmbeddedLanguageCompletionProvider : CompletionProvider
+    /// <typeparam name="T">the <see cref="IEmbeddedLanguage"/> relevant to the completion provder.</typeparam>
+    internal abstract class AbstractEmbeddedLanguageCompletionProvider<T> : LSPCompletionProvider where T : class, IEmbeddedLanguage
     {
         public const string EmbeddedProviderName = "EmbeddedProvider";
 
-        private ImmutableArray<IEmbeddedLanguage> _languageProviders;
+        private Lazy<T?>? _lazyLanguageFeature;
```

Comment:
Would love to hear a better way to do this - basically I'm trying to capture three states here - not yet initialized at all vs initialized with null value vs initialized with non-null value